test/stdlib_checks.dart

```diff
@@ -63,6 +63,8 @@ Object? deepToJson(Object? object) {
       result = object.map((x) => deepToJson(x)).toList();
     case Map() when object.keys.every((k) => k is String):
       result = object.map((k, v) => MapEntry<String, dynamic>(k, deepToJson(v)));
+    case Map() when object.keys.every((k) => k is int):
+      result = object.map((k, v) => MapEntry<String, dynamic>(k.toString(), deepToJson(v)));
```

Comment:
Hmm interesting.

The idea in the existing code here is that if you have a Map whose keys aren't strings, that is not a valid JSON-like value — not a value that a `toJson` method should be emitting. Keys in actual JSON objects are strings (just like in JS objects).

E.g. see the `deepToJson` doc:
```
/// All JSON atoms (numbers, booleans, null, and strings) are used directly.
/// All JSON containers (lists, and maps with string keys) are copied
/// as their elements are converted recursively.
/// For any other value, a dynamic call `.toJson()` is made and
/// should return either a JSON atom or a JSON container.
Object? deepToJson(Object? object) {
```

Does `jsonEncode` accept a value like this? If it does, we should probably accommodate it here too.

If not, but `jsonEncode` gives a more helpful error message about this, then it'd be good to try to give an equally helpful error.

Reply:
In DartPad, this fails:

```dart
import 'dart:convert';

main() {
  print(jsonEncode({ 1: 123 }));
}
```

with

```quote
Uncaught Error: Converting object to an encodable object failed: Instance of 'JsLinkedHashMap<int, int>'
```

Reply:
Compare that with the output of this test on `main` (256be9aa375f40bba21768f67cb8851dd98474d0):

```dart
    test('', () {
      check({1: 1}).jsonEquals({1: 1});
    });
```

```
$ flutter test test/model/unreads_test.dart 
00:02 +0 -1: constructor  [E]                                                      
  Converting object to an encodable object failed: _Map len:1
  test/stdlib_checks.dart 49:5        deepToJson
  test/stdlib_checks.dart 79:26       JsonChecks.jsonEquals
  test/model/unreads_test.dart 98:21  main.<fn>.<fn>
```

Reply:
That's a `JsonUnsupportedObjectError` that we throw in `deepToJson`:

```dart
  final Object? shallowlyConverted;
  try {
    shallowlyConverted = (object as dynamic).toJson();
  } catch (e) {
    throw JsonUnsupportedObjectError(object, cause: e);
  }
```


Inspecting its `cause` with `print(e)` gives:

```dart
NoSuchMethodError: Class '_Map<int, int>' has no instance method 'toJson'.
Receiver: _Map len:1
Tried calling: toJson()
```

Reply:
Maybe it would help to make that `cause` appear in the test failure output? If the test writer already knows not to expect `deepToJson` to convert an int-keyed map to a string-keyed map (as JS's `JSON.stringify` does), then that output is likely to be a helpful clue, because it points to an int-keyed map.

Reply:
It looks like we give the same error as `jsonEncode` does here. It's different in DartPad vs. `flutter test` because the former is running Dart by compiling to JS; but in `flutter test`, the `jsonEncode` version gives the same "_Map len:1" stringification.

Which isn't a coincidence, because as the implementation comment says:
```dart
Object? deepToJson(Object? object) {
  // Implementation is based on the recursion underlying [jsonEncode],
  // at [_JsonStringifier.writeObject] in the stdlib's convert/json.dart .
  // (We leave out the cycle-checking, for simplicity / out of laziness.)
```

and that includes largely throwing the same exceptions.

I'm inclined to not try to optimize the error messages beyond what `jsonEncode` gives. For example it's not clear that this would be helpful:

> If the test writer already knows not to expect `deepToJson` to convert an int-keyed map to a string-keyed map (as JS's `JSON.stringify` does), then that output is likely to be a helpful clue,

because that doesn't seem to have been the scenario.